src/libcollections/binary_heap.rs

```diff
@@ -554,6 +553,70 @@ impl<T: Ord> BinaryHeap<T> {
     pub fn clear(&mut self) { self.drain(); }
 }
 
+/// Hole represents a hole in a slice i.e. an index without valid value
+/// (because it was moved from or duplicated).
+/// In drop, `Hole` will restore the slice by filling the hole
+/// position with the value that was originally removed.
+struct Hole<'a, T: 'a> {
+    data: &'a mut [T],
+    elt: Option<T>,
+    pos: usize,
+}
+
+impl<'a, T> Hole<'a, T> {
+    /// Create a new Hole at index `pos`.
+    pub fn new(data: &'a mut [T], pos: usize) -> Self {
```

Comment:
Idiomatically code in the rest of the standard library does not do this, so let's stick to existing conventions. We can add `pub` if necessary at a later date, but code should be as conservative as possible in exports today.